service/r2/store/kv/store.go

```diff
@@ -366,28 +373,59 @@ func (s *store) DeleteRollupRule(
 func (s *store) FetchRollupRuleHistory(namespaceID, rollupRuleID string) ([]*models.RollupRuleView, error) {
 	rs, err := s.ruleStore.ReadRuleSet(namespaceID)
 	if err != nil {
-		return nil, s.handleUpstreamError(err)
+		return nil, handleUpstreamError(err)
 	}
 
 	rrs, err := rs.RollupRules()
 	if err != nil {
-		return nil, s.handleUpstreamError(err)
+		return nil, handleUpstreamError(err)
 	}
 
 	hist, exists := rrs[rollupRuleID]
 	if !exists {
-		return nil, s.rollupRuleNotFoundError(namespaceID, rollupRuleID)
+		return nil, rollupRuleNotFoundError(namespaceID, rollupRuleID)
 	}
 	return hist, nil
 }
 
 func (s *store) Close() { s.ruleStore.Close() }
 
-func (s *store) newUpdateMeta(uOpts r2.UpdateOptions) rules.UpdateMetadata {
+func (s *store) UpdateRuleSet(
+	rsChanges changes.RuleSetChanges,
+	version int,
+	uOpts r2store.UpdateOptions,
+) (*models.RuleSet, error) {
+	rs, err := s.FetchRuleSet(rsChanges.Namespace)
+	if err != nil {
+		return nil, handleUpstreamError(err)
+	}
+	if version != rs.Version() {
+		return nil, r2.NewConflictError("mismatched ruleset version number")
+	}
+
+	mutable := rs.ToMutableRuleSet().Clone()
+	err = mutable.ApplyRuleSetChanges(rsChanges, s.newUpdateMeta(uOpts))
+	if err != nil {
+		return nil, handleUpstreamError(err)
+	}
+	err = s.ruleStore.WriteRuleSet(mutable)
+	if err != nil {
+		return nil, handleUpstreamError(err)
+	}
+
+	snapshot, err := s.FetchRuleSetSnapshot(string(rs.Namespace()))
```

Comment:
nit: you can use `rsChanges.Namespace` to avoid a `[]byte` to `string` conversion here.

Also conventionally we return the snapshot view (e.g., `FetchRuleSet` returns `*models.RuleSetSnapshotView`) so we should be consistent. You can always do the conversion to ruleset on the client side if needed.